### PR TITLE
river-status: add keyboard_layout event

### PR DESCRIPTION
**Pull request by MaxVerevkin**
I'm new to zig and server-side wayland, so please let me know if if there are better ways to do certain things.

`sway` was used as a reference.

- Simple test client: https://github.com/MaxVerevkin/river-kbd-layout-watcher
- Status bar integration: https://github.com/MaxVerevkin/i3status-rust/tree/river-kbd

### CODE REVIEW

**Comment by MaxVerevkin** [comment]
Test client: https://github.com/MaxVerevkin/rust-river-kbd-layout-watcher

**Review by ifreund — COMMENTED** [review]
I haven't looked into this in great depth yet, but I don't see how this handles two keyboards on the same seat with different active layouts...

**Comment by MaxVerevkin** [comment]
Hmm, I guess the event should also contain some kind of input identifier?

File: river/SeatStatus.zig [hunk 1]
```diff
@@ -78,3 +87,22 @@ pub fn sendMode(self: Self, mode: [*:0]const u8) void {
...
+    if (self.seat_status.getVersion() < 5) return;
+
+    const identifier = std.fmt.allocPrint(
+        util.gpa,
+        "{s}\x00",
+        .{device.identifier},
+    ) catch return;
+    defer util.gpa.free(identifier);
+
+    const c_ident = @ptrCast([*:0]const u8, identifier.ptr);
```
Comment (MaxVerevkin) [inline river/SeatStatus.zig:101]: Should I change `InputDevice.identifier` to be a null terminated sting to avoid this?

**Comment by MaxVerevkin** [comment]
Should this event be sent when keyboard is created / destroyed?

**Comment by MaxVerevkin** [comment]
An alternative to this can be a separate protocol extension that exposes input information similar to how wl_output/xdg_output describes outputs. Are there any such extensions already (I couldn't find any)? Any opinions on this?

File: river/SeatStatus.zig (hunk 1 above)
Comment (ifreund) [reply river/SeatStatus.zig:101]: Sure, that would be more idiomatic. Also note that allocPrintZ() can be used instead of this pointer cast.

**Comment by Leon-Plickat** [comment]
> Edit: I suppose this is covered by https://github.com/riverwm/river/issues/71

I don't think so. My idea was to limit that protocol to devices handled by libinput, because unifying multiple API surfaces into a single protocol seems a bit unreasonable.

**Comment by MaxVerevkin** [comment]
Right now the event is sent when on binding, on keyboard creation and when layout changes. I think it's better so send it when keyboard is destroyed too. Should I add `keyboard_destroyed` event or just send `keyboard_layout_clear`?

**Comment by ifreund** [comment]
I don't think I'm going to accept this feature as part of a river-specific protocol. Other compositors have the exact same issue and it would be better for this to be solved in a compositor-agnostic way if possible.

See for example https://github.com/labwc/labwc/issues/1317

Alternatively, with my plans for a complete window management protocol this information may end up being exposed to the window manager process by river which could then relay it to status bars and the like.

Sorry to keep you hanging on this for so long without a clear decision one way or the other.

**Comment by MaxVerevkin** [comment]
> Other compositors have the exact same issue and it would be better for this to be solved in a compositor-agnostic way if possible.

I absolutely agree.

In the mean time, for anyone interested, I will continue maintaining this branch, until a better solution is designed.

**Comment by karampok** [comment]
Hi, I am interested, as I do miss a such feature (just be able to print in the bar the current layout).
Do you know if there there anyother  way (standalone tool) that I can query the layout in wayland/river? 
xdb-switch status or setxkbmap -query or localectl do not work. I only have seen sway-lock to be able to do that.

Thanks!

**Pull request closed without merging**